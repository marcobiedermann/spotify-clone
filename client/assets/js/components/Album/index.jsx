import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import './style.css';

const Album = ({
  artists, id, images, name,
}) => (
  <figure className="album">
    {images[1] && (
      <Link to={`/albums/${id}`}>
        <img src={images[1].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <h3>
        <Link to={`/albums/${id}`}>
          {name}
        </Link>
      </h3>
      {artists[0] && (
        <h4>
          <Link to={`/artists/${artists[0].id}`}>
            {artists[0].name}
          </Link>
        </h4>
      )}
    </figcaption>
  </figure>
);

Album.propTypes = {
  artists: PropTypes.arrayOf(PropTypes.shape()),
  id: PropTypes.string,
  images: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
};

Album.defaultProps = {
  artists: [],
  id: '',
  images: [],
  name: '',
};

export default Album;
