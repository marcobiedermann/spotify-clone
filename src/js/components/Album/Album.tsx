import React from 'react';
import { Link } from 'react-router-dom';
import AlbumTracks from '../AlbumTracks';
import Image from '../Image';

const defaultImageSize = 100;

interface Artist {
  id?: string;
  name?: string;
}

interface Image {
  height: number | null;
  url: string;
  width: number | null;
}

interface Item {
  artists?: Artist[];
  duration_ms?: number;
  id?: string;
  name?: string;
  track_number?: number;
}

interface Tracks {
  items: Item[];
}

export interface AlbumProps {
  artists?: Artist[];
  id: string;
  images: Image[];
  name: string;
  tracks?: Tracks;
}

function Album(props: AlbumProps): JSX.Element {
  const { artists = [], id, images, name, tracks } = props;
  const artist = artists[0];
  const image = images[0];

  return (
    <div>
      <figure>
        {image && (
          <Link to={`/albums/${id}`}>
            <Image
              {...image}
              alt={name}
              width={image.width || defaultImageSize}
              height={image.height || defaultImageSize}
            />
          </Link>
        )}
        <figcaption>
          <h3>
            <Link to={`/albums/${id}`}>{name}</Link>
          </h3>
          <h4>
            <Link to={`/artists/${artist.id}`}>{artist.name}</Link>
          </h4>
        </figcaption>
      </figure>
      {tracks && <AlbumTracks items={tracks.items} />}
    </div>
  );
}

export default Album;
