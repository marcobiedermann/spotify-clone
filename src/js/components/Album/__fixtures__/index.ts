const album = {
  artists: [
    {
      id: '0TnOYISbd1XYRBk9myaseg',
      name: 'Pitbull',
    },
  ],
  id: '4aawyAB9vmqN3uQ7FjRGTy',
  images: [
    {
      height: 640,
      url: 'https://i.scdn.co/image/ab67616d0000b2732c5b24ecfa39523a75c993c4',
      width: 640,
    },
  ],
  name: 'Global Warming',
  tracks: {
    items: [
      {
        artists: [
          {
            id: '0TnOYISbd1XYRBk9myaseg',
            name: 'Pitbull',
          },
        ],
        duration_ms: 85400,
        id: '6OmhkSOpvYBokMKQxpIGx2',
        name: 'Global Warming (feat. Sensato)',
        track_number: 1,
      },
    ],
  },
};

export default album;
