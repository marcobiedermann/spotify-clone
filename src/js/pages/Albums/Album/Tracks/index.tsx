import { Helmet } from 'react-helmet';

function TracksPage(): JSX.Element {
  return (
    <>
      <Helmet>
        <title>Tracks</title>
      </Helmet>
      Tracks
    </>
  );
}

export default TracksPage;
