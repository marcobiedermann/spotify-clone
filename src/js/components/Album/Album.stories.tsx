import type { Meta, StoryObj } from '@storybook/react';
import { MemoryRouter } from 'react-router-dom';
import Album from '.';
import albumFixtures from './__fixtures__';

const meta: Meta<typeof Album> = {
  component: Album,
  args: albumFixtures,
  decorators: [
    (Story) => (
      <MemoryRouter>
        <Story />
      </MemoryRouter>
    ),
  ],
};

type Story = StoryObj<typeof Album>;

const Default: Story = {};

export { Default };
export default meta;
