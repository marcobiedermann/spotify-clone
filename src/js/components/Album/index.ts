import Album from './Album';

export default Album;
export * from './Album';
