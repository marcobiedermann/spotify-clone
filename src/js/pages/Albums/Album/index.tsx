import React from 'react';
import { Helmet } from 'react-helmet';
import { useParams } from 'react-router-dom';
import { Album, Error, Loader } from '../../../components';
import { useAlbum } from '../../../hooks/albums';

function AlbumPage(): JSX.Element {
  const { albumId } = useParams();
  const { data, error, isError, isLoading } = useAlbum(albumId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isLoading) {
    return <Loader />;
  }

  const { name } = data;

  return (
    <>
      <Helmet>
        <title>{name}</title>
      </Helmet>
      <Album {...data} />
    </>
  );
}

export default AlbumPage;
