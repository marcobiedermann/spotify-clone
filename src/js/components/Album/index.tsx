import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import Image from '../Image';
import Tracks from '../Tracks';

interface Artist {
  id: string;
  name: string;
}

interface Image {
  height: number;
  url: string;
  width: number;
}

interface Item {
  duration_ms: number;
  id: string;
  name: string;
  track_number: number;
}

interface Tracks {
  items: Item[];
}

export interface AlbumProps {
  artists: Artist[];
  id: string;
  images: Image[];
  name: string;
  tracks?: Tracks;
}

const Album: FC<AlbumProps> = (props) => {
  const { artists, id, images, name, tracks } = props;

  return (
    <div>
      <figure>
        {images[1] && (
          <Link to={`/albums/${id}`}>
            <Image {...images[1]} alt={name} />
          </Link>
        )}
        <figcaption>
          <h3>
            <Link to={`/albums/${id}`}>{name}</Link>
          </h3>
          {artists[0] && (
            <h4>
              <Link to={`/artists/${artists[0].id}`}>{artists[0].name}</Link>
            </h4>
          )}
        </figcaption>
      </figure>
      {tracks && <Tracks items={tracks.items} />}
    </div>
  );
};

export default Album;
