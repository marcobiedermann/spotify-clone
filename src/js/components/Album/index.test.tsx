import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import Album from '.';
import albumFixtures from './__fixtures__';

describe('Album component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Album {...albumFixtures} />
      </MemoryRouter>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
