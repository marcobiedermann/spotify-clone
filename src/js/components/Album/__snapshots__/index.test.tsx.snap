// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album component renders correctly 1`] = `
<div>
  <figure>
    <figcaption>
      <h3>
        <a
          href="/albums/"
        />
      </h3>
      <h4>
        <a
          href="/artists/"
        />
      </h4>
    </figcaption>
  </figure>
  <table>
    <tbody>
      <tr>
        <td>
          100
        </td>
        <td />
        <td>
          00:00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
